src/tqec/position.py

```diff
@@ -60,3 +63,59 @@ class Displacement:
 
     x: int
     y: int
+
+
+@dataclass(frozen=True, order=True)
+class Position3D:
+    """A 3D integer position."""
+
+    x: int
+    y: int
+    z: int
+
+    def __post_init__(self) -> None:
+        if any(not isinstance(i, int) for i in astuple(self)):
+            raise TQECException("Position must be an integer.")
+
+    def shift_by(self, dx: int = 0, dy: int = 0, dz: int = 0) -> "Position3D":
+        """Shift the position by the given offset."""
+        return Position3D(self.x + dx, self.y + dy, self.z + dz)
+
+    def is_neighbour(self, other: "Position3D") -> bool:
+        """Check if the other position is near to this position, i.e. Manhattan distance is 1."""
+        return (
+            abs(self.x - other.x) + abs(self.y - other.y) + abs(self.z - other.z) == 1
+        )
+
+    def as_tuple(self) -> tuple[int, int, int]:
+        """Return the position as a tuple."""
+        return astuple(self)
+
+    def __str__(self) -> str:
+        return f"({self.x},{self.y},{self.z})"
+
+
+class Direction3D(Enum):
+    """Axis directions in the 3D spacetime diagram."""
+
+    X = auto()
+    Y = auto()
+    Z = auto()
+
+    @staticmethod
+    def all() -> list["Direction3D"]:
+        """Get all directions."""
+        return [Direction3D.X, Direction3D.Y, Direction3D.Z]
+
+    @staticmethod
+    def from_axis_index(i: int) -> "Direction3D":
```

Comment:
Due to the very limited valid inputs, I would rather annotate with `typing.Literal`. If you import `typing`, the following should be stricter and provide more information to the user:
```suggestion
    def from_axis_index(i: typing.Literal[0, 1, 2]) -> "Direction3D":
```
If you prefer to keep `int` as typing, I would then add a runtime check and a `TQECException` with a nice explanatory message about why the entry should not be above 2.

Reply:
I often come across some type issues with `Literal` when I want to pass a `int`/`string` as the argument of the function. For example, the following code should raise a type error when using int literal for the argument type:
```py
a = [1, 3, 2]
i = a.index(3)
Direction3D.from_axis_index(i)
```
What's the general solution to this without tedious `cast` everywhere?

Reply:
The solution I would use is to keep the annotation generic (`int` as initially), but introduce a runtime check in the method:
```suggestion
    def from_axis_index(i: int) -> "Direction3D":
        if not 0 <= i <= 2:
            raise TQECException(
                f"Could not get the axis related to index {i}. Axes "
                f"are indexed 0, 1 or 2 and 0 <= {i} <= 2 should be verified."
            )
```
I do not have a suitable solution to make the typing information right, this is up to the type checkers, and they do not implement that fully.